src/rust/cryptography-x509-validation/src/lib.rs

```diff
@@ -11,6 +11,323 @@ pub mod policy;
+                                return Err(ValidationError::Other(
+                                    "excluded name constraint matched SAN".into(),
+                                ));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn build_chain_inner(
+        &self,
+        working_cert: &'a Certificate<'chain>,
+        current_depth: u8,
+        is_leaf: bool,
+        extensions: &'a Extensions<'chain>,
+    ) -> Result<IntermediateChain<'chain>, ValidationError> {
+        if current_depth > self.policy.max_chain_depth {
+            return Err(ValidationError::Other(
+                "chain construction exceeds max depth".into(),
+            ));
+        }
+
+        // Look in the store's root set to see if the working cert is listed.
+        // If it is, we've reached the end.
+        if self.store.contains(working_cert) {
+            let mut constraints = vec![];
+            self.build_name_constraints(&mut constraints, extensions)?;
+            return Ok((vec![working_cert.clone()], constraints));
+        }
+
+        // Otherwise, we collect a list of potential issuers for this cert,
+        // and continue with the first that verifies.
+        let mut last_err: Option<ValidationError> = None;
+        for issuing_cert_candidate in self.potential_issuers(working_cert) {
+            // A candidate issuer is said to verify if it both
+            // signs for the working certificate and conforms to the
+            // policy.
+            let issuer_extensions = issuing_cert_candidate.extensions()?;
+            match self.policy.valid_issuer(
+                issuing_cert_candidate,
+                working_cert,
+                current_depth,
+                &issuer_extensions,
+            ) {
+                Ok(next_depth) => {
+                    match self.build_chain_inner(
+                        issuing_cert_candidate,
+                        next_depth,
+                        false,
+                        &issuer_extensions,
+                    ) {
+                        Ok((remaining, mut constraints)) => {
+                            // Name constraints are not applied to self-issued certificates unless they're
+                            // the leaf certificate in the chain.
+                            //
+                            // NOTE: We can't simply check the `current_depth` since self-issued
+                            // certificates don't increase the working depth.
+                            let skip_name_constraints =
+                                cert_is_self_issued(working_cert) && !is_leaf;
+                            if skip_name_constraints
+                                || self
+                                    .apply_name_constraints(&constraints, extensions)
+                                    .is_ok()
+                            {
```

Comment:
That comes from this language in RFC 5280:

>    Name constraints are not applied to self-issued certificates (unless
   the certificate is the final certificate in the path).  (This could
   prevent CAs that use name constraints from employing self-issued
   certificates to implement key rollover.)

I can update the comment to cite that directly.

Reply:
b6de1f91c